sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -451,30 +468,24 @@ case class HashShuffleSpec(
       false
   }
 
-  override def canCreatePartitioning: Boolean = true
+  override def canCreatePartitioning: Boolean = {
+    // To avoid potential data skew, we don't allow `HashShuffleSpec` to create partitioning if
+    // the hash partition keys are not the full join keys (the cluster keys). Then the planner
+    // will add shuffles with the default partitioning of `ClusteredDistribution`, which uses all
+    // the join keys.
+    if (SQLConf.get.getConf(SQLConf.REQUIRE_ALL_JOIN_KEYS_AS_PARTITION_KEYS)) {
+      distribution.clustering.forall(x => partitioning.expressions.exists(_.semanticEquals(x)))
```

Comment:
Do we need to require `partitioning.expressions` to be exactly same with `distribution.clustering` as well? e.g. for followed cases:

```
partitioning.expressions: [a, b]
distribution.clustering: [b, a]
```

```
partitioning.expressions: [a, b, a]
distribution.clustering: [a, b]
```

Reply:
Good point. To fully restore to the previous behavior, we should require an exact match, though I think the current change should cover the data skew issues.

I'll make the change to be conservative.

Reply:
Yes the case where partitioning expressions is a superset of distribution clustering should already be rejected by `HashPartitioning#satisfies`, but it maybe better to make it more explicit here.